src/main/java/bio/terra/service/filedata/flight/ingest/IngestBuildAndWriteScratchLoadFileStep.java

```diff
@@ -0,0 +1,110 @@
+import bio.terra.service.filedata.flight.FileMapKeys;
+import bio.terra.service.filedata.google.gcs.GcsPdao;
+import bio.terra.service.resourcemanagement.google.GoogleBucketResource;
+import bio.terra.service.snapshot.exception.CorruptMetadataException;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.StepResult;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class IngestBuildAndWriteScratchLoadFileStep extends SkippableStep {
+  private final ObjectMapper objectMapper;
+  private final GcsPdao gcsPdao;
+  private final Dataset dataset;
+
+  public IngestBuildAndWriteScratchLoadFileStep(
+      ObjectMapper objectMapper,
+      GcsPdao gcsPdao,
+      Dataset dataset,
+      Predicate<FlightContext> skipCondition) {
+    super(skipCondition);
+    this.objectMapper = objectMapper;
+    this.gcsPdao = gcsPdao;
+    this.dataset = dataset;
+  }
+
+  @Override
+  public StepResult doSkippableStep(FlightContext context) {
+    var workingMap = context.getWorkingMap();
+    IngestRequestModel ingestRequest = IngestUtils.getIngestRequestModel(context);
+
+    List<String> errors = new ArrayList<>();
+    try (Stream<JsonNode> jsonNodes =
+        IngestUtils.getJsonNodesStreamFromFile(
+            gcsPdao, objectMapper, ingestRequest, dataset, errors)) {
+
+      List<String> fileColumns =
+          workingMap.get(IngestMapKeys.TABLE_SCHEMA_FILE_COLUMNS, List.class);
+      BulkLoadArrayResultModel result =
+          workingMap.get(IngestMapKeys.BULK_LOAD_RESULT, BulkLoadArrayResultModel.class);
+
+      // Part 1 -> Build the src-target hash to file id Map
+      Map<Integer, String> pathToFileIdMap =
+          result.getLoadFileResults().stream()
+              .collect(
+                  Collectors.toMap(
+                      model -> Objects.hash(model.getSourcePath(), model.getTargetPath()),
+                      BulkLoadFileResultModel::getFileId));
+
+      // Part 2 -> Replace BulkLoadFileModels with file id
+      Stream<String> linesWithFileIds =
+          jsonNodes
+              .peek(
+                  node -> {
+                    for (var columnName : fileColumns) {
+                      JsonNode fileRefNode = node.get(columnName);
+                      if (fileRefNode != null && fileRefNode.isObject()) {
+                        // replace
+                        BulkLoadFileModel fileModel =
+                            Objects.requireNonNull(
+                                objectMapper.convertValue(fileRefNode, BulkLoadFileModel.class));
+                        int fileKey =
+                            Objects.hash(fileModel.getSourcePath(), fileModel.getTargetPath());
```

Comment:
Instead of repeating the `Objects.hash()` call maybe a static method on this class would be better, to make it obvious that the same hash function is used in both cases. It would be best to add it to `BulkLoadFileModel` but since that's auto-generated this is probably the best we can do.

Reply:
I'm not sure its much use to make a static method for each of the types described above.